rfcs/2022-12-09-541-reload-api.md

```diff
@@ -0,0 +1,277 @@
+significant maintenance burden and it's unlikely that we want to continue
+investing in it for Vector's use case. By starting a new, simpler API
+foundation, we can set the stage to migrate current GraphQL API use cases
+towards technology more suited to the team maintaining it.
+
+### Make providers more synchronous
+
+Likely the smallest change we could make would be to allow a feedback mechanism
+for providers. This would allow us to implement the control server as a
+provider with fewer structural changes to Vector's main run loop. Concretely,
+this could look something like adding an optional oneshot response channel to
+the `SignalTo::ReloadFromConfigBuilder` variant used by providers.
+
+We don't take this route because the signal-based focus of reloading is already
+overloaded enough to make the process difficult to follow. It's not a good
+match for what we're trying to do and would like further obscure the logic,
+making bugs more difficult to track down. In fact, given the limited uptake of
+the provider system in general, it's possible that it should be reworked to fit
+the same pattern as the new control server. This could be done as future work.
+
+### More isolated control socket mode
+
+Another option would be to branch earlier in `Application::run`, make options
+related to the control server exclusive with normal options like `--config`,
+and have more of a separate mode of operation for this functionality. This
+would remove some concerns about potential strange interactions between
+multiple paths for reloading (e.g. watching a file while running a control
+server while running an HTTP provider) and potentially require less refactoring
+work than what's proposed in this RFC.
+
+This approach wasn't chosen because the divergence for "normal" Vector
+operation would itself be a path for potentially confusing differences in
+behavior. It also would not move us closer to an improved overall design for
+this portion of Vector, and would create another path that we would later
+likely want to unify.
+
+### gRPC instead of JSON
+
+Instead of a REST-ish JSON API, we could use gRPC. We already use the
+technology in other areas of Vector (native source and sink), so it wouldn't
+have the same downside as GraphQL, while providing some of the benefits of an
+IDL.
+
+This approach was not chosen for a couple of reasons:
+- Protobuf adds complexity (i.e. code generation) to any client implementation.
+- The ability to interact with the API from simple tools like cURL is lost.
+- The only current endpoint accepts a Vector config, for which we already have
+  a JSON Schema and we would not want to translate to Protobuf.
+- Many of the benefits of gRPC are around high performance, load balancing,
+  health checking, etc, none of which is relevant to a small local control API.
+
+## Outstanding Questions
+
+- What is the expected flow for setting global options at startup (especially
```

Comment:
My current understanding is that, in general, OPW will send Vector two configuration files on start-up: the user-provided configuration file as-is and a generated configuration file containing enterprise components. 

> Use normal file loading with only global options set?

Does this mean it'd be preferable to have `global` options extracted into a separate file? I'm trying to understand if this is a technical concern or if it's moreso a UX concern to make it clear that `global` options don't reload. 

And would this separate global config be explicit to the user (e.g. `--global-config` option on OPW) or would we handle it internally (e.g. extract any `global` section from the `--config` file)? 

cc: @leebenson for some of these questions.

Reply:
> My current understanding is that, in general, OPW will send Vector two configuration files on start-up: the user-provided configuration file as-is and a generated configuration file containing enterprise components.

Yes, that is what happens now, except for the "sending" part. Currently we're just passing through a user-provided config file, so the `--config` path passed to OPW is then also passed to Vector (after some validation). The enterprise components are written to a second file by OPW and that is passed as an additional file to Vector. Currently, the split into two files is so that we don't need to modify the user-provided file. The issues with that largely go away with an API, but we likely still need to have a "bootstrap" configuration file of global options that is passed to Vector at startup.

> Does this mean it'd be preferable to have global options extracted into a separate file? I'm trying to understand if this is a technical concern or if it's moreso a UX concern to make it clear that global options don't reload.

It's a bit of both. There is a technical limitation in that certain things must be set at startup and therefore aren't able to be passed in via a reload API. Then there is the UX issue of making it clear which things those are and logically separating them from things like sources/sinks/transforms so that the user is aware of the difference.

> And would this separate global config be explicit to the user (e.g. --global-config option on OPW) or would we handle it internally (e.g. extract any global section from the --config file)?

Ideally I think that most users should not need to worry about global config. We can have defaults that work well for most use cases, and for use cases where they do need set we can have a separate mechanism that makes it clear they are not reloadable (e.g. ENV vars or command line flags). So we can essentially remove them from the config schema that we allow to be passed to Vector.

Reply:
My own thinking is pretty much as @lukesteensen described.

Currently in OP UI, we insert an `[enterprise]` block for bootstrapping Vector.

By the time this functionality lands, that block should be redundant. API keys will be set by env vars (and/or perhaps CLI flags, if that's still on the cards). 

Since Vector will be a managed process, other 'global' config, like the `[api]` section or Vector paths, will be determined either by the deployment method or by runtime flags to OPW. There should be little, if any, global config left for a user to configure in UI.

One side effect to this is that the config we're reporting back to OP will also likely be devoid of these sections.

Reply:
> Yes, that is what happens now, except for the "sending" part. Currently we're just passing through a user-provided config file

Sorry, "send" was bad phrasing on my part -- pass through is what I intended.

> Ideally I think that most users should not need to worry about global config. We can have defaults that work well for most use cases, and for use cases where they do need set we can have a separate mechanism that makes it clear they are not reloadable (e.g. ENV vars or command line flags). So we can essentially remove them from the config schema that we allow to be passed to Vector.

Ok, sounds good. What I suggested seems unnecessary then.

> One side effect to this is that the config we're reporting back to OP will also likely be devoid of these sections.

Right. As you mention, if none of this `global` configuration is intended to be in the UI, it seems like this is fine then.

Thanks for the clarifications.